### PR TITLE
Add cypress tests (navbar, userpanel)

### PR DESCRIPTION
**Pull request by anshulahuja98**
Work towards #4153

### CODE REVIEW

**Comment by anshulahuja98** [comment]
@eloyekunle pl have a look

**Comment by anshulahuja98** [comment]
PR ready to be reviewed

**Comment by floreks** [comment]
I'll review when I have some time. Probably on monday.

**Comment by floreks** [comment]
First and most important create PO (Page Objects) to encapsulate all view-related stuff.

Example
```typescript
export class MembersPage {
  static getAddMemberBtn(): Cypress.Chainable<any> {
    return cy.get('#add-member-top-btn');
  }

  static getEditMemberDialogGroupCombobox(): Cypress.Chainable<any> {
    return cy.get('#edit-member-dialog-group-combobox');
  }

  static getEditMemberDialogSaveBtn(): Cypress.Chainable<any> {
    return cy.get('#edit-member-dialog-edit-btn');
  }

  static getTable(): Cypress.Chainable<any> {
    return cy.get('tbody');
  }

...

  static getTableRow(email: string): Cypress.Chainable<any> {
    return this.getTableRowEmailColumn(email).parent();
  }

  static getTableRowGroupColumn(email: string): Cypress.Chainable<any> {
    return this.getTableRow(email).find('td.mat-column-group');
  }
...
```

**Comment by anshulahuja98** [comment]
> First and most important create PO (Page Objects) to encapsulate all view-related stuff.
> 
> Example
> 
> ```ts
> export class MembersPage {
>   static getAddMemberBtn(): Cypress.Chainable<any> {
>     return cy.get('#add-member-top-btn');
>   }
> 
>   static getEditMemberDialogGroupCombobox(): Cypress.Chainable<any> {
>     return cy.get('#edit-member-dialog-group-combobox');
>   }
> 
>   static getEditMemberDialogSaveBtn(): Cypress.Chainable<any> {
>     return cy.get('#edit-member-dialog-edit-btn');
>   }
> 
>   static getTable(): Cypress.Chainable<any> {
>     return cy.get('tbody');
>   }
> 
> ...
> 
>   static getTableRow(email: string): Cypress.Chainable<any> {
>     return this.getTableRowEmailColumn(email).parent();
>   }
> 
>   static getTableRowGroupColumn(email: string): Cypress.Chainable<any> {
>     return this.getTableRow(email).find('td.mat-column-group');
>   }
> ...
> ```

Got the point
Can you perhaps point to some repo which I can refer to for optimum directory structures and in general other structuring tips?

**Comment by floreks** [comment]
> Can you perhaps point to some repo which I can refer to for optimum directory structures and in general other structuring tips for these tests and POs?

Unfortunately, I haven't found any good public repo I can recommend. Our internal repo where we use cypress is private.

Few tips:
- Create PO class per view and put it under `cypress/pages/xxx.po.ts`
- It's nice to export stuff like Conditions, `contain`, `not.contain` to an enum. This can be put under `cypress/utils/condition.ts`.
- Try to write actual tests `it` block so `cy.xxx` will not be there directly. There might be some exceptions, however usually it should use page objects.

Examples
```ts
  it('should login', () => {
    login(email, password);

    cy.url().should(Condition.Include, 'projects');
  });
```

```ts
  it('should go to members view', () => {
    MembersPage.visit();
  });
```

**Comment by anshulahuja98** [comment]
> > Can you perhaps point to some repo which I can refer to for optimum directory structures and in general other structuring tips for these tests and POs?
> 
> Unfortunately, I haven't found any good public repo I can recommend. Our internal repo where we use cypress is private.
> 
> Few tips:
> 
> * Create PO class per view and put it under `cypress/pages/xxx.po.ts`
> * It's nice to export stuff like Conditions, `contain`, `not.contain` to an enum. This can be put under `cypress/utils/condition.ts`.
> * Try to write actual tests `it` block so `cy.xxx` will not be there directly. There might be some exceptions (assertions), however usually it should use page objects.
> 
> Examples
> 
> ```ts
>   it('should login', () => {
>     login(email, password);
>     ProjectsPage.verifyUrl();
>   });
> ```
> 
> ```ts
>   it('should go to members view', () => {
>     MembersPage.visit();
>     MembersPage.verifyUrl();
>   });
> ```
 
And where do I put the actual tests?

**Comment by floreks** [comment]
> And where do I put the actual tests?

It depends on the test. If it's user story then it can go under `cypress/integration/stories`. If it's a regular set of tests, then `cypress/integration/specs`.

> I couldn't find an implementation in ts as mentioned in your example.

I have posted it above. You simply create a class with static methods.

**Comment by anshulahuja98** [comment]
> > I couldn't find an implementation in ts as mentioned in your example.
> 
> I have posted it above. You simply create a class with static methods.

Actually I meant in terms of the required imports etc. And other common files

So, from what I could find online, it would require a typescript description file.
And probably some other stuff.

**Comment by floreks** [comment]
You do not need any kind of definition files.

To the `cypress/plugins` dir add `cy-ts-preprocessor.js`
```js
const wp = require('@cypress/webpack-preprocessor');

const webpackOptions = {
  resolve: {extensions: ['.ts', '.js']},
  module: {
    rules: [{
      test: /\.ts$/,
      exclude: [/node_modules/],
      use: [{
        loader: 'ts-loader',
        options: {configFile: 'cypress/tsconfig.e2e.json'}
      }]
    }]
  }
};

const options = { webpackOptions };
module.exports = wp(options);
```

and `index.js`

```js
const cypressTypeScriptPreprocessor = require('./cy-ts-preprocessor');

module.exports = on => {
  on('file:preprocessor', cypressTypeScriptPreprocessor);
};
```

to `cypress` dir add `tsconfig.e2e.json`
```json
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "strict": true,
    "baseUrl": "../node_modules",
    "target": "es5",
    "lib": ["es5", "dom"],
    "types": ["cypress"]
  },
  "include": [
    "integration/**/*.ts",
    "pages/**/*.ts",
    "utils/*.ts",
    "support/*.ts",
    "../node_modules/cypress/**/*.ts"]
}
```

That should be it. Cypress should then understand typescript files in other directories.

**Comment by anshulahuja98** [comment]
Thank You @floreks !
I'll get working on it.

**Comment by atoato88** [comment]
/uncc
I'm a reviewer in `i18n/ja/` only.
Sorry that I cannot review js code on this PR.

**Comment by maciaszczykm** [comment]
Are you willing to finish it?

**Comment by anshulahuja98** [comment]
I have been stuck on an issue doing this, I pinged @floreks on slack regarding this
Haven't heard from him

If anyone else could assist me
I would be glad to close this asap

**Comment by floreks** [comment]
I'd have to look into the code and try to configure it on my own. Don't have any ideas right now why it fails.

**Comment by anshulahuja98** [comment]
Hey @floreks I finally got the tests to work. 

Can you please have a look. I have added some tests for now. I'll push some more commits with more tests soon.

**Comment by anshulahuja98** [comment]
PR ready for review

File: cypress/integration/specs/nav.spec.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+describe('Navbar', () => {
+  before(() => {
+    NavbarPage.visitHome();
+  });
+  describe('Navbar Cluster Items', () => {
+    it('cluster', () => {
+      NavbarPage.clickItem('#nav-cluster');
+      NavbarPage.assertUrlContains('cluster');
+    });
```
Comment (maciaszczykm) [inline cypress/integration/specs/nav.spec.ts]: nit: Can you separate tests with new lines?

File: cypress/integration/specs/nav.spec.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
+import {NavbarPage} from '../../pages/navbarPage';
+
+describe('Navbar', () => {
+  before(() => {
+    NavbarPage.visitHome();
+  });
+  describe('Navbar Cluster Items', () => {
+    it('cluster', () => {
+      NavbarPage.clickItem('#nav-cluster');
```
Comment (maciaszczykm) [inline cypress/integration/specs/nav.spec.ts]: It would be very good to have IDs like #nav-cluster in the PO (page object) file.

Comment (anshulahuja98) [reply cypress/integration/specs/nav.spec.ts]: Made the required changes.

File: cypress/integration/specs/nav.spec.ts (hunk 1 above)
Comment (anshulahuja98) [reply cypress/integration/specs/nav.spec.ts]: Made the required changes.

**Comment by anshulahuja98** [comment]
@maciaszczykm let me know if there are any more changes required, once it's sufficient, I'll fix the merge conflicts.

**Comment by anshulahuja98** [comment]
@maciaszczykm @floreks

**Comment by maciaszczykm** [comment]
Sorry for not responding for a while, can you resolve the conflicts?

**Comment by anshulahuja98** [comment]
@maciaszczykm resolved the conflicts

**Comment by anshulahuja98** [comment]
@maciaszczykm @floreks

File: cypress/integration/specs/userpanel.spec.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import {UserpanelPage} from '../../pages/userpanelPage';
+
+describe('Userpanel', () => {
+  before(() => {
+    UserpanelPage.visitHome();
+  });
+
+  describe('Basic tests', () => {
```
Comment (maciaszczykm) [inline cypress/integration/specs/userpanel.spec.ts]: There is no need for this category as it it single within user panel.

File: cypress/integration/specs/nav.spec.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
+import {NavbarPage} from '../../pages/navbarPage';
+
+describe('Navbar', () => {
+  before(() => {
+    NavbarPage.visitHome();
+  });
+
+  describe('Navbar Cluster Items', () => {
```
Comment (maciaszczykm) [inline cypress/integration/specs/nav.spec.ts]: This categorization is good, I'd only remove the `Navbar` from `Navbar Cluster Items` to not repeat it.

**Comment by maciaszczykm** [comment]
Looks good overall, please fix these 2 minor things and then we can merge. We can improve it incrementally. Thanks for patience :)

**Comment by maciaszczykm** [comment]
You need to resolve then conflicts then I will approve :)

**Comment by anshulahuja98** [comment]
> You need to resolve then conflicts then I will approve :)

Yes, I was just working on the rebase.

**Comment by anshulahuja98** [comment]
> You need to resolve then conflicts then I will approve :)

Resolved

**Comment by anshulahuja98** [comment]
Travis failing is not due to this PR, it has something to do with the last 4-5 PRs merged in the repo. And my PR is rebased upto them
@maciaszczykm

**Comment by shu-mutou** [comment]
Could you rebase again? `package.json` seems to be reverted. Thanks for your pacience.

File: package.json [hunk 5]
```diff
@@ -103,96 +99,98 @@
...
     "sockjs-client": "1.4.0",
-    "systemjs": "6.1.9",
-    "tslib": "1.10.0",
+    "systemjs": "6.1.7",
+    "tslib": "^1.9.0",
     "web-animations-js": "2.3.2",
     "xterm": "3.14.5",
-    "zone.js": "0.9.1"
+    "zone.js": "0.9.1",
+    "@cypress/webpack-preprocessor": "latest"
```
Comment (shu-mutou) [inline package.json]: this line should be added in "devDependencies" instead "dependencies".

**Comment by anshulahuja98** [comment]
I forgot to add this dependency(been a while) while manually updating package.json , hope I didn't miss anything else 😅 

Although, I think the other part will still fail due to the current build failing 
https://github.com/kubernetes/dashboard/commit/b5c3e9f65d50db550ee385e43c0d5be741144849

**Comment by anshulahuja98** [comment]
@shu-mutou @maciaszczykm the build is finally fixed now!

**Comment by anshulahuja98** [comment]
By the time the PR gets approved, new merge conflicts keep arising from just previously merged PRs bumping dependencies.
🤕

**Comment by maciaszczykm** [comment]
There is likely some issue in one of the merged JSON files. Please verify.

**Comment by anshulahuja98** [comment]
@maciaszczykm  need to remove the do-not-merge label

**Pull request merged**